src/inmanta/ast/statements/generator.py

```diff
@@ -268,7 +267,151 @@ def nested_blocks(self) -> Iterator["BasicBlock"]:
         yield self.module
 
 
-class If(ExpressionStatement):
+# TODO: include general flow in docstring (here or requires_emit?): requires_emit -> resume -> execute
+class ListComprehension(RawResumer, ExpressionStatement):
+    """
+    A list comprehension expression, e.g. `["hello {{world}}" for world in worlds]`.
+    """
+
+    __slots__ = ("loop_var", "value_expression", "iterable")
+
+    def __init__(self, value_expression: ExpressionStatement, loop_var: LocatableString, iterable: ExpressionStatement) -> None:
+        super().__init__()
+        self.value_expression: ExpressionStatement = value_expression
+        self.loop_var: LocatableString = loop_var
+        self.iterable: ExpressionStatement = iterable
+        # TODO: extend anchors?
+
+    def normalize(self, *, lhs_attribute: Optional[AttributeAssignmentLHS] = None) -> None:
+        self.value_expression.normalize(lhs_attribute=lhs_attribute)
+        self.iterable.normalize()
+
+    def requires(self) -> list[str]:
+        # exclude loop var, unless it shadows an occurrence in iterable
+        return list(set(self.iterable.requires()) - {self.loop_var} + set(self.value_expression.requires()))
+
+    def requires_emit(
```

Comment:
it feels a bit unnatural for this implementation to be in `requires_emit` and not in `requires_emit_gradual`

Reply:
I wasn't sure myself which direction felt more natural. This one is an extension to `requires_emit` to optionally make it gradual, the other direction is an extension to `requires_emit_gradual` to optionally make it not gradual. Both felt unnatural to me with this one the lesser of two evils because `requires_emit_gradual` explicitly has `gradual` in the name so dropping it felt even more weird. But I'm willing to swap it if you prefer. Or if you see a third approach that does feel natural I'm all for that.